### PR TITLE
Added mso_schema_site_contract_service_graph_listener resource to manage Azure CNC - Listeners and Rules

### PR DESCRIPTION
**Pull request by sajagana**
Note:

Added mso_schema_site_contract_service_graph_listener resource to manage Azure CNC Contract Service Graph Load Balancer - Listeners and Rules

### CODE REVIEW

File: website/docs/r/schema_site_contract_service_graph_listener.html.markdown [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+* `schema_id` - (Required) The schema ID of the Listener.
+* `template_name` - (Required) The template name of the Listener.
+* `contract_name` - (Required) The contract name of the Listener.
+* `site_id` - (Required) The site ID under which the Listener is deployed.
+* `service_node_index` - (Required) The service node index of the Service Graph under which the Listener is deployed.
+* `listener_name` - (Required) The name of the Listener.
+* `protocol` - (Required) The protocol of the Listener. Allowed values are `http`, `https`, `tcp`, `udp`, `tls` and `inherit`.
+* `port` - (Required) The port number of the Listener.
+* `security_policy` - (Optional) The security policy of the Listener. Allowed values are `default`, `eLBSecurityPolicy-2016-08`, `eLBSecurityPolicy-FS-2018-06`, `eLBSecurityPolicy-TLS-1-2-2017-01`, `eLBSecurityPolicy-TLS-1-2-Ext-2018-06`, `eLBSecurityPolicy-TLS-1-1-2017-01`, `eLBSecurityPolicy-2015-05`, `eLBSecurityPolicy-TLS-1-0-2015-04`, `AppGwSslPolicyDefault`, `AppGwSslPolicy20150501`, `AppGwSslPolicy20170401` and `AppGwSslPolicy20170401S`.
```
Comment (akinross) [inline website/docs/r/schema_site_contract_service_graph_listener.html.markdown:122]: For all the optional input with allowed values, is there default behaviour when not provided?

File: website/docs/r/schema_site_contract_service_graph_listener.html.markdown [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+* `frontend_ip_dn` - (Optional) The frontend IP DN of the Cloud L4-L7 Network Load Balancer device. The frontend IP should be configured in the Azure Cloud Network Controller. The frontend IP configuration option will be available under the `Advanced Settings` of the Network Load Balancer device.
+* `ssl_certificates` - (Optional) The SSL Certificates information of the Listener.
+  * `name` - (Required) The key ring name of the SSL Certificate.
+  * `target_dn` - (Required) The key ring DN of the SSL Certificate.
+  * `default` - (Read-Only) The default boolean flag of the SSL Certificate.
```
Comment (akinross) [inline website/docs/r/schema_site_contract_service_graph_listener.html.markdown:127]: Should this be read-only?

File: mso/resource_mso_schema_site_contract_service_graph_listener.go [hunk 3]
```diff
@@ -0,0 +1,918 @@
...
+			},
+			"port": &schema.Schema{
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+			"security_policy": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ValidateFunc: validation.StringInSlice([]string{
```
Comment (akinross) [inline mso/resource_mso_schema_site_contract_service_graph_listener.go:83]: Did you consider using a simpler input format and mapping table?

File: mso/resource_mso_schema_site_contract_service_graph_listener.go [hunk 4]
```diff
@@ -0,0 +1,918 @@
...
+								"forward",
+								"redirect",
+								"haPort",
+							}, false),
+						},
+						"content_type": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+							Computed: true,
+							ValidateFunc: validation.StringInSlice([]string{
```
Comment (akinross) [inline mso/resource_mso_schema_site_contract_service_graph_listener.go:196]: Did you consider using a simpler input format and mapping table?

File: mso/resource_mso_schema_site_contract_service_graph_listener.go [hunk 5]
```diff
@@ -0,0 +1,918 @@
...
+						"condition": &schema.Schema{
+							Type:         schema.TypeString,
+							Optional:     true,
+							Computed:     true,
+							ValidateFunc: validation.StringLenBetween(1, 1000),
+						},
+						"action_type": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+							ValidateFunc: validation.StringInSlice([]string{
```
Comment (akinross) [inline mso/resource_mso_schema_site_contract_service_graph_listener.go:185]: Do we want to use camel case input? Did you consider using a simpler input format and mapping table?

File: mso/resource_mso_schema_site_contract_service_graph_listener.go [hunk 6]
```diff
@@ -0,0 +1,918 @@
...
+						"redirect_port": &schema.Schema{
+							Type:     schema.TypeInt,
+							Optional: true,
+							Computed: true,
+						},
+						"redirect_code": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+							Computed: true,
+							ValidateFunc: validation.StringInSlice([]string{
```
Comment (akinross) [inline mso/resource_mso_schema_site_contract_service_graph_listener.go:319]: Do we want to use camel case input? Did you consider using a simpler input format and mapping table?

File: mso/resource_mso_schema_site_contract_service_graph_listener.go [hunk 7]
```diff
@@ -0,0 +1,918 @@
...
+							}, false),
+						},
+						"port": &schema.Schema{
+							Type:     schema.TypeInt,
+							Required: true,
+						},
+						"protocol": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+							ValidateFunc: validation.StringInSlice([]string{
```
Comment (akinross) [inline mso/resource_mso_schema_site_contract_service_graph_listener.go:211]: should we define once and reuse the protocol inputs?

File: mso/resource_mso_schema_site_contract_service_graph_listener.go [hunk 8]
```diff
@@ -0,0 +1,918 @@
...
+									"unhealthy_threshold": &schema.Schema{
+										Type:     schema.TypeInt,
+										Optional: true,
+										Computed: true,
+									},
+									"use_host_from_rule": &schema.Schema{
+										Type:     schema.TypeString,
+										Optional: true,
+										Computed: true,
+										ValidateFunc: validation.StringInSlice([]string{
```
Comment (akinross) [inline mso/resource_mso_schema_site_contract_service_graph_listener.go:377]: Should we use boolean input here?

File: mso/resource_mso_schema_site_contract_service_graph_listener.go [hunk 9]
```diff
@@ -0,0 +1,918 @@
...
+								"unspecified",
+								"primary",
+								"secondary",
+							}, false),
+						},
+					},
+				},
+			},
+		},
+		CustomizeDiff: func(diff *schema.ResourceDiff, v interface{}) error {
```
Comment (akinross) [inline mso/resource_mso_schema_site_contract_service_graph_listener.go:411]: Can you comment why the customizeDiff is required? Think that would make it easier for people to understand.

**Review by akinross — CHANGES_REQUESTED** [review]

File: website/docs/r/schema_site_contract_service_graph_listener.html.markdown (hunk 1 above)
Comment (sajagana) [reply website/docs/r/schema_site_contract_service_graph_listener.html.markdown]: The MSO/NDO setting it to `null` when the `security_policy` is not provided during the creation. That is not a valid option, So ignored the default value here.

File: website/docs/r/schema_site_contract_service_graph_listener.html.markdown (hunk 2 above)
Comment (sajagana) [reply website/docs/r/schema_site_contract_service_graph_listener.html.markdown]: As per the offline discussion, removed from the schema.

File: website/docs/r/schema_site_contract_service_graph_listener.html.markdown [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+  * `redirect_host_name` - (Optional) The redirect host name of the Rule.
+  * `redirect_path` - (Optional) The redirect path of the Rule.
+  * `redirect_query` - (Optional) The redirect query of the Rule.
+  * `response_code` - (Optional) The response code of the Rule.
+  * `response_body` - (Optional) The response body of the Rule.
+  * `redirect_protocol` - (Optional) The redirect protocol of the Rule. Allowed values are `http`, `https`, `tcp`, `udp`, `tls` and `inherit`. The MSO/NDO defaults to `inherit` when unset during creation.
+  * `redirect_port` - (Optional) The redirect port of the Rule.
+  * `redirect_code` - (Optional) The redirect code of the Rule. Allowed values are `unknown`, `permanently_moved`, `found`, `see_other` and `temporary_redirect`. The MSO/NDO defaults to `permanently_moved` when unset during creation.
+  * `target_ip_type` - (Optional) The target IP type of the Rule. Allowed values are `unspecified`, `primary` and `secondary`.  The MSO/NDO defaults to `unspecified` when unset during creation.
+  * `health_check` - (Optional) The Health Checks information of the Listener Rule.
```
Comment (akinross) [inline website/docs/r/schema_site_contract_service_graph_listener.html.markdown:150]: To keep consistent with the others of the Rule?
```suggestion
  * `health_check` - (Optional) The Health Checks information of the Rule.
```

File: website/docs/r/schema_site_contract_service_graph_listener.html.markdown [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+    * `path` - (Optional) The path of the Health Checks. The MSO/NDO defaults to `` when unset during creation.
+    * `interval` - (Optional) The interval(seconds) of the Health Checks. The MSO/NDO defaults to `0` when unset during creation.
+    * `timeout` - (Optional) The timeout(seconds) of the Health Checks. The MSO/NDO defaults to `0` when unset during creation.
+    * `unhealthy_threshold` - (Optional) The unhealthy threshold of the Health Checks. The MSO/NDO defaults to `0` when unset during creation.
+    * `use_host_from_rule` - (Optional) The use host from rule of the Health Checks. Allowed values are `true` and `false`. The MSO/NDO defaults to `false` when unset during creation.
+    * `success_code` - (Optional) The success code (code range) of the Health Checks. The MSO/NDO defaults to `200-399` when unset during creation.
+    * `host` - (Optional) The host of the Health Checks. The `host` attribute will be enabled when the `use_host_from_rule` is `false`.
+  * `provider_epg_ref` - (Optional) The Provider EPG information of the Listener Rule.
```
Comment (akinross) [inline website/docs/r/schema_site_contract_service_graph_listener.html.markdown:160]: To keep consistent with the others of the Rule?
```suggestion
  * `provider_epg_ref` - (Optional) The Provider EPG information of the Rule.
```

File: website/docs/d/schema_site_contract_service_graph_listener.html.markdown [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+  * `redirect_host_name` - (Read-Only) The redirect host name of the Rule.
+  * `redirect_path` - (Read-Only) The redirect path of the Rule.
+  * `redirect_query` - (Read-Only) The redirect query of the Rule.
+  * `response_code` - (Read-Only) The response code of the Rule.
+  * `response_body` - (Read-Only) The response body of the Rule.
+  * `redirect_protocol` - (Read-Only) The redirect protocol of the Rule.
+  * `redirect_port` - (Read-Only) The redirect port of the Rule.
+  * `redirect_code` - (Read-Only) The redirect code of the Rule.
+  * `target_ip_type` - (Read-Only) The target IP type of the Rule.
+  * `health_check` - (Read-Only) The Health Checks information of the Listener Rule.
```
Comment (akinross) [inline website/docs/d/schema_site_contract_service_graph_listener.html.markdown:67]: To keep consistent with the others of the Rule?
```suggestion
  * `health_check` - (Read-Only) The Health Checks information of the Rule.
```

File: website/docs/d/schema_site_contract_service_graph_listener.html.markdown [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    * `port` - (Read-Only) The port number of the Health Checks.
+    * `protocol` - (Read-Only) The port number of the Health Checks.
+    * `path` - (Read-Only) The path of the Health Checks.
+    * `interval` - (Read-Only) The interval(seconds) of the Health Checks.
+    * `timeout` - (Read-Only) The timeout(seconds) of the Health Checks.
+    * `unhealthy_threshold` - (Read-Only) The unhealthy threshold of the Health Checks.
+    * `use_host_from_rule` - (Read-Only) The use host from rule of the Health Checks.
+    * `success_code` - (Read-Only) The success code (code range) of the Health Checks.
+    * `host` - (Read-Only) The host of the Health Checks.
+  * `provider_epg_ref` - (Read-Only) The Provider EPG information of the Listener Rule.
```
Comment (akinross) [inline website/docs/d/schema_site_contract_service_graph_listener.html.markdown:77]: To keep consistent with the others of the Rule?
```suggestion
  * `provider_epg_ref` - (Read-Only) The Provider EPG information of the Rule.
```

**Review by akinross — CHANGES_REQUESTED** [review]

File: website/docs/r/schema_site_contract_service_graph_listener.html.markdown [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+
+# Note: #
+This resource is only compatible with NDO versions 3.7 and 4.2+. NDO versions 4.0 and 4.1 are not supported.
+
+## Example Usage ##
+
+```hcl
+
+# Sample Listener configuration for the Application Load Balancer
+resource "mso_schema_site_contract_service_graph_listener" "application_load_balancer_node1" {
```
[the remark this answers is not in the document]
Comment (akinross) [reply website/docs/r/schema_site_contract_service_graph_listener.html.markdown]: now both are named the same example, should be example_1 and example_2 ( or example_<name> )

**Review by akinross — APPROVED** [review]
LGTM

**Review by anvitha-jain — DISMISSED** [review]

**Review by samiib — APPROVED** [review]
LGTM

**Review by lhercot — CHANGES_REQUESTED** [review]

**Review by akinross — APPROVED** [review]
LGTM

**Review by samiib — APPROVED** [review]
LGTM

**Review by shrsr — APPROVED** [review]
LGTM

**Review by anvitha-jain — APPROVED** [review]

**Review by lhercot — APPROVED** [review]
LGTM

**Pull request merged**